File: solutions/scala/difference-of-squares/1/src/main/scala/DifferenceOfSquares.scala
object DifferenceOfSquares {

  def differenceOfSquares(number: Int): Int =
    this.squareOfSum(number) - this.sumOfSquares(number)

  @scala.annotation.tailrec
  def sumOfSquares(number: Int, accumulated: Int = 0): Int =
    if (number > 0) {
      this.sumOfSquares(number - 1, accumulated + number * number)
    }
    else {
      accumulated
    }

  @scala.annotation.tailrec
  def squareOfSum(number: Int, accumulated: Int = 0): Int =
    if (number > 0) {
      this.squareOfSum(number - 1, accumulated + number)
    }
    else {
      accumulated * accumulated
    }
}

File: solutions/erlang/difference-of-squares/1/src/difference_of_squares.erl
-module(difference_of_squares).

-export([difference_of_squares/1, square_of_sum/1, sum_of_squares/1]).

difference_of_squares(Number) ->
  square_of_sum(Number) - sum_of_squares(Number).

square_of_sum(Number) ->
  square_of_sum(Number, 0).

square_of_sum(Number, Accumulated) when Number > 0 ->
  square_of_sum(Number - 1, Accumulated + Number);

square_of_sum(_Number, Accumulated) ->
  Accumulated * Accumulated.

sum_of_squares(Number) ->
  sum_of_squares(Number, 0).

sum_of_squares(Number, Accumulated) when Number > 0 ->
  sum_of_squares(Number - 1, Accumulated + Number * Number);

sum_of_squares(_Number, Accumulated) ->
  Accumulated.
